packages/react-test-renderer/src/ReactShallowRenderer.js

```diff
@@ -99,6 +118,259 @@ class ReactShallowRenderer {
+              newState = reducer(newState, action);
+              update = update.next;
+            } while (update !== null);
+
+            this._workInProgressHook.memoizedState = newState;
+
+            return [newState, dispatch];
+          }
+        }
+        return [this._workInProgressHook.memoizedState, dispatch];
+      } else {
+        if (reducer === basicStateReducer) {
+          // Special case for `useState`.
+          if (typeof initialState === 'function') {
+            initialState = initialState();
+          }
+        } else if (initialAction !== undefined && initialAction !== null) {
+          initialState = reducer(initialState, initialAction);
+        }
+        this._workInProgressHook.memoizedState = initialState;
+        const queue: UpdateQueue<A> = (this._workInProgressHook.queue = {
+          last: null,
+          dispatch: null,
+        });
+        const dispatch: Dispatch<
+          A,
+        > = (queue.dispatch = (this._dispatchAction.bind(
+          this,
+          this._currentlyRenderingComponent,
+          queue,
+        ): any));
+        return [this._workInProgressHook.memoizedState, dispatch];
+      }
+    };
+
+    const useState = <S>(
+      initialState: (() => S) | S,
+    ): [S, Dispatch<BasicStateAction<S>>] => {
+      return useReducer(
+        basicStateReducer,
+        // useReducer has a special case to support lazy useState initializers
+        (initialState: any),
+      );
+    };
+
+    const useMemo = <T>(
+      nextCreate: () => T,
+      inputs: Array<mixed> | void | null,
+    ): T => {
+      this._validateCurrentlyRenderingComponent();
+      this._createWorkInProgressHook();
+
+      const nextInputs =
+        inputs !== undefined && inputs !== null ? inputs : [nextCreate];
+
+      if (
+        this._workInProgressHook !== null &&
+        this._workInProgressHook.memoizedState !== null
+      ) {
+        const prevState = this._workInProgressHook.memoizedState;
+        const prevInputs = prevState[1];
+        if (areHookInputsEqual(nextInputs, prevInputs)) {
+          return prevState[0];
+        }
+      }
+
+      const nextValue = nextCreate();
+      this._workInProgressHook.memoizedState = [nextValue, nextInputs];
+      return nextValue;
+    };
+
+    const useRef = <T>(initialValue: T): {current: T} => {
+      this._validateCurrentlyRenderingComponent();
+      this._createWorkInProgressHook();
+      const previousRef = this._workInProgressHook.memoizedState;
+      if (previousRef === null) {
+        const ref = {current: initialValue};
+        if (__DEV__) {
+          Object.seal(ref);
+        }
+        this._workInProgressHook.memoizedState = ref;
+        return ref;
+      } else {
+        return previousRef;
+      }
+    };
+
+    const readContext = <T>(
+      context: ReactContext<T>,
+      observedBits: void | number | boolean,
+    ): T => {
+      return context._currentValue;
```

Comment:
The partial renderer impl varies from yours:
```js
function readContext<T>(
  context: ReactContext<T>,
  observedBits: void | number | boolean,
): T {
  let threadID = currentThreadID;
  validateContextBounds(context, threadID);
  return context[threadID];
}
```

Not sure of this reason for this. Is it a mistake?

Reply:
We use threads on the SSR version, but I believed we didn't for the DOM version. I'll take another look at this.

Reply:
I don't think this will be relevant to shallow.

Reply:
Agreed. Just didn't have any context on the SSR case for this and wanted verification that it wouldn't cause any unforeseen trouble here.